File: py/src/pypaginate/dataset.py
"""Resident in-memory dataset: marshal once, query many times.

Marshals a sequence into the Rust core ONCE, then answers repeated filter / sort
/ search / page queries natively (the core returns indices; this wrapper selects
your original objects). Build once, query many — the only in-memory shape where
crossing into Rust pays off versus native ``list`` operations.
"""

from __future__ import annotations

from collections.abc import Sequence
from typing import Generic, TypeVar

from pypaginate import _core, _native
from pypaginate.errors import FilterError, SearchError, SortError
from pypaginate.pages import OffsetPage
from pypaginate.params import OffsetParams
from pypaginate.specs import FilterSpec, SearchSpec, SortSpec


ItemT = TypeVar("ItemT")


class Dataset(Generic[ItemT]):
    """An in-memory dataset queried natively (marshalled once)."""

    __slots__ = ("_inner", "_items")

    def __init__(self, items: Sequence[ItemT]) -> None:
        self._items: list[ItemT] = list(items)
        self._inner = _core.Dataset(self._items)

    def __len__(self) -> int:
        return len(self._items)

    def __repr__(self) -> str:
        return f"Dataset({len(self._items)} items)"

    def _select(self, indices: list[int]) -> list[ItemT]:
        return [self._items[i] for i in indices]

    def filter(self, filters: Sequence[FilterSpec]) -> list[ItemT]:
        """Rows matching the flat ``filters`` (each keeps its own AND/OR logic)."""
        tuples = _native.filter_tuples(filters)
        try:
            return self._select(self._inner.filter(tuples))
        except (KeyError, ValueError) as exc:
            raise FilterError(str(exc)) from exc

    def sort(self, sorting: Sequence[SortSpec]) -> list[ItemT]:
        """Rows sorted by ``sorting`` (stable, null-aware)."""
        tuples = _native.sort_tuples(sorting)
        try:
            return self._select(self._inner.sort(tuples))
        except (KeyError, ValueError) as exc:
            raise SortError(str(exc)) from exc

    def search(self, spec: SearchSpec) -> list[ItemT]:
        """Rows ranked by relevance of ``spec`` over its fields."""
        try:
            return self._select(self._search(spec))
        except (KeyError, ValueError) as exc:
            raise SearchError(str(exc)) from exc

    def _search(self, spec: SearchSpec) -> list[int]:
        return self._inner.search(
            spec.query,
            list(spec.fields),
            spec.mode or "contains",
            spec.fuzzy or "exact",
            spec.threshold if spec.threshold is not None else 30,
            spec.min_length if spec.min_length is not None else 1,
            spec.max_results,
        )

    def page(
        self,
        params: OffsetParams,
        *,
        filters: Sequence[FilterSpec] | None = None,
        sorting: Sequence[SortSpec] | None = None,
        search: SearchSpec | None = None,
    ) -> OffsetPage[ItemT]:
        """Filter + search + sort + offset-paginate in one native call."""
        result = self._inner.page(
            params.page,
            params.limit,
            _native.filter_tuples(filters or []),
            _native.sort_tuples(sorting or []),
            _native.search_stage(search) if search is not None else None,
        )
        items = self._select(result["indices"])
        return _native.build_offset_page(items, result["total"], params)


__all__ = ["Dataset"]

File: ts/src/dataset.ts
/**
 * A resident dataset: marshal rows into the Rust core ONCE, then run many
 * filter / sort / search / page queries natively (the core returns indices, this
 * wrapper maps them back to your own objects). This is the only in-memory shape
 * where crossing into Rust pays off for JS — the one-shot `*Indices` helpers
 * re-marshal the whole array on every call. Build once, query many times.
 *
 * Core engine failures are re-typed into the typed `FilterError` / `SortError` /
 * `SearchError`, matching pypaginate's exception taxonomy.
 */
import * as core from "@cyblow/paginate-core";

import { FilterError, PaginateError, SearchError, SortError, rethrowEngineError } from "./errors.js";
import type { OffsetParams } from "./params.js";
import type { OffsetPage } from "./pages.js";
import type { FilterSpec, SearchSpec, SortSpec } from "./specs.js";

export class Dataset<T extends object> {
  private readonly inner: core.Dataset;
  private readonly rows: readonly T[];

  constructor(rows: readonly T[]) {
    this.rows = rows;
    this.inner = new core.Dataset(rows as unknown[]);
  }

  /** Number of rows held. */
  get size(): number {
    return this.inner.size;
  }

  /** Rows matching the filter specs (pypaginate's exact semantics). */
  filter(specs: readonly FilterSpec[]): T[] {
    try {
      return this.select(this.inner.filter(specs as unknown as core.FilterSpecInput[]));
    } catch (e) {
      rethrowEngineError(e, FilterError);
    }
  }

  /** Rows sorted by the specs (null-aware, stable). */
  sort(specs: readonly SortSpec[]): T[] {
    try {
      return this.select(this.inner.sort(specs as unknown as core.SortSpecInput[]));
    } catch (e) {
      rethrowEngineError(e, SortError);
    }
  }

  /** Rows ranked by relevance of the search spec's query over its fields. */
  search(spec: SearchSpec): T[] {
    try {
      return this.select(
        this.inner.search(
          spec.query,
          spec.fields as string[],
          spec.mode,
          spec.fuzzy,
          spec.threshold,
          spec.minLength,
          spec.maxResults,
          spec.weights as Record<string, number> | undefined,
        ),
      );
    } catch (e) {
      rethrowEngineError(e, SearchError);
    }
  }

  /**
   * Filter + search + sort + offset-paginate in ONE native call. `search` is a
   * match-filter (keep rows matching the query); an explicit `sorting` still
   * decides the order, and the resident trigram index prunes fuzzy candidates.
   */
  page(
    params: OffsetParams,
    opts: {
      filters?: readonly FilterSpec[];
      sorting?: readonly SortSpec[];
      search?: SearchSpec;
    } = {},
  ): OffsetPage<T> {
    try {
      const result = this.inner.page(
        params.page,
        params.limit,
        opts.filters as unknown as core.FilterSpecInput[] | undefined,
        opts.sorting as unknown as core.SortSpecInput[] | undefined,
        opts.search ? searchStageArg(opts.search) : undefined,
      );
      return {
        items: this.select(result.indices),
        total: Number(result.total),
        page: Number(result.page),
        pages: Number(result.pages),
        limit: params.limit,
        hasNext: result.hasNext,
        hasPrevious: result.hasPrevious,
      };
    } catch (e) {
      // Multi-op: the core's message prefix selects filter/sort/search; anything
      // else (incl. field-not-found, which carries no op) → the base error.
      rethrowEngineError(e, PaginateError);
    }
  }

  private select(indices: number[]): T[] {
    return indices.map((i) => this.rows[i] as T);
  }
}

/** The `{ query, fields, mode, fuzzy, threshold }` object the core page search
 * stage parses (a match-filter; `minLength` / `maxResults` do not apply). */
function searchStageArg(spec: SearchSpec): Record<string, unknown> {
  return {
    query: spec.query,
    fields: spec.fields,
    mode: spec.mode,
    fuzzy: spec.fuzzy,
    threshold: spec.threshold,
  };
}
